rtlive/sources/data_de.py

```diff
@@ -289,8 +303,81 @@ def forecast_DE(df: pandas.DataFrame):
+    
+    Parameters
+    ----------
+    daily_data : pandas.Series
+        The known daily test counts. This will be used to determine at which point the predicted data starts.
+    total_tests_reported : pandas.Series
+        Series containing test count totals. Is expected to contain NaN gaps in the data. The Differences between
+        two reports will be used to make sure the total number of tests in the prediction matches this data.
+    predicted_daily_data: pandas.Series
+        The predicted data that will be scaled so the number of tests between two reports matches total_tests_reported.
+
+    Returns
+    -------
+    corrected_prediction: pandas.Series
+        The scaled prediction
+    """
+    day = pandas.Timedelta('1D')
+    
+    first_unknown_daily_data = max(daily_data[~daily_data.isna()].index) + day
+    
+    report_data = total_tests_reported[~total_tests_reported.isna()].diff()
+    
+    #find the start date of the week report that contaisn this date:
+    report_index = report_data.index.get_loc(first_unknown_daily_data,method="ffill")
+    
+    #We assume, that the report for a day includes the data for the day itself.
+    first_report_date = report_data.index[report_index] + day
+    
+    corrected_prediction = predicted_daily_data.copy()
+
+    for last_report_date, reportsum in report_data[first_report_date:].items():
+        if numpy.isnan(reportsum):
+            break
+        if first_unknown_daily_data > first_report_date:
+            # We have daily data for parts of this report. Subtract the known days from the report sum and only scale the unknown data
+            reportsum -= daily_data[first_report_date:first_unknown_daily_data - day].sum()
+            first_report_date = first_unknown_daily_data
+        corrected_prediction[first_report_date:last_report_date] *= reportsum/predicted_daily_data[first_report_date:last_report_date].sum()
+        first_report_date = last_report_date + day
+    return corrected_prediction
+
+def estimate_test_percentages_for_regions(df: pandas.DataFrame):
+    """ Estimate the distribution of tests across the regions in the DataFrame. Uses the last week for which daily new_test data is available for all regions.
+    
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        The dataframe containing the new_test column with a [region, date] index as genereated by get_testcounts_DE. An `all` region has to be included.
+        
+    Returns
+    -------
+    region_test_percentages: pandas.DataFrame
+        Dataframe with region-index and percentage values between 0 and 1 for each region.
+    """
+    # Find the last date for each region for which we have data and get the first date of those. So the last date for which he have date for all regions.
+    last_date_with_data_for_all_regions = df.new_tests[~df.new_tests.isna()].groupby('region').tail(1).reset_index()['date'].min()
```

Comment:
I assume there is a better way to achieve this, but I couldn't come up with one. Especially the `tail(1).reset_index()` feels/looks weird.

Reply:
Yes, we also tried to find a better way, but ended up with the same as you did.